### PR TITLE
Feature/contacts

### PR DESCRIPTION
**Pull request by SunnyCat2013**

### CODE REVIEW

File: react-redux/contactsMVC/components/Conversation.jsx [hunk 1]
```diff
@@ -0,0 +1,103 @@
+import React, { PropTypes } from 'react';
+import ReactDOM from 'react-dom';
+import { Card, CardText } from 'material-ui/Card';
+import { List } from 'material-ui/List';
+import Subheader from 'material-ui/Subheader';
+import TextField from 'material-ui/TextField';
+import FlatButton from 'material-ui/FlatButton';
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/Conversation.jsx]:
``` javascript
import { Card, CardText } from 'material-ui/Card';
import { List } from 'material-ui/List';
import React, { PropTypes } from 'react';
import ReactDOM from 'react-dom';
import Subheader from 'material-ui/Subheader';
import TextField from 'material-ui/TextField';
import FlatButton from 'material-ui/FlatButton';
```

带括号的先(specific import)，不带的后(default import)。仅仅为了好看一些

File: react-redux/contactsMVC/components/Conversation.jsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+import request from 'superagent';
+import promisity from 'superagent-as-promised';
+promisity(request);
+
+import CommentBox from './commentbox'
+
+let nextCommentId = 0;
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/Conversation.jsx]: don't use global id, use `node-uuid`

Comment (ZhengHe-MD) [reply react-redux/contactsMVC/components/Conversation.jsx]: generage uuid for each comment

``` sh
npm install node-uuid --save
```

``` javascript
import uuid from 'node-uuid';
const nextCommentId = uuid.v4();
```

File: react-redux/contactsMVC/components/Conversation.jsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+class Conversation extends React.Component {
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/Conversation.jsx]: you can also

``` javascript
import React, { Component, PropTypes } from 'react';
class Conversation extends Component {}
```

File: react-redux/contactsMVC/components/Conversation.jsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    this.onSaveComment = this.onSaveComment.bind(this);
+    this.getRandomUser = this.getRandomUser.bind(this);
+    this.changeInputValue = this.changeInputValue.bind(this);
+
+    this.state = {
+      inputValue: ''
+    }
+  }
+
+  changeInputValue (e) {
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/Conversation.jsx]: name all your event handler as `onDoSomething`.
in this specific case, use `onChangeInputValue`

Comment (ZhengHe-MD) [reply react-redux/contactsMVC/components/Conversation.jsx]: or you can use `onSomethingDone`
in this specific case, use `onInputValueChanged`
pick whichever you like

File: react-redux/contactsMVC/components/Conversation.jsx [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    this.setState({ inputValue: e.target.value });
+  }
+
+  getShowComments(showComments) {
+    return (
+      <div>
+        {
+          showComments.map(commentObject => (
+            <CommentBox key={commentObject.id} commentObject={commentObject} store={this.props.store} />
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/Conversation.jsx]: do this when there are more than two props

``` javascript
<CommentBox
  key={commentObject.id}
  commentObject={commentObject}
  store={this.props.store}
/>
```

File: react-redux/contactsMVC/components/Conversation.jsx [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+          ))
+        }
+      </div>
+    );
+  }
+
+  render() {
+    const show_comments = this.props.store.getState();
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/Conversation.jsx]: rename all your **show_comments** to **showComments**

File: react-redux/contactsMVC/components/Conversation.jsx [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    console.log('render');
+    return (
+        <Card className="sidebar">
+          <List>
+            <Subheader>Conversation</Subheader>
+            {
+              show_comments.length > 0 ? this.getShowComments(show_comments) : <div><h1>There is no comment.</h1></div>
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/Conversation.jsx]:
``` javascript
{
  showComments.length > 0 ?
    this.getShowComments(showComments) :
    <div><h1>There is not comment.</h1></div>
}
```

or 

``` javascript
{
  showComments.length > 0
    ? this.getShowComments(showComments)
    : <div><h1>There is not comment.</h1></div>
}
```

File: react-redux/contactsMVC/components/Conversation.jsx [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            }
+          </List>
+          <CardText>
+            <TextField
+              floatingLabelText="Comment"
+              value={this.state.inputValue}
+              onChange={this.changeInputValue}
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/Conversation.jsx]: event handler naming convention

File: react-redux/contactsMVC/components/Conversation.jsx [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  getRandomUser() {
+    request
+      .get('https://randomuser.me/api/')
+      .set('Accept', 'application/json')
+      .then(results => {
+        const data = JSON.parse(results.text).results[0];
+        this.onSaveComment(data.login.username, data.picture.thumbnail);
+      }).catch(error => console.error);
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/Conversation.jsx]: put **.catch** to a new line

File: react-redux/contactsMVC/components/Conversation.jsx [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  onSaveComment(
+    userName,
+    userThumbnail,
+  ) {
+    console.log(this.state.inputValue);
+
+    console.log(userName);
+    console.log(userThumbnail);
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/Conversation.jsx]: remove all your console.log

File: react-redux/contactsMVC/components/commentbox.jsx [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+import ReactDOM from 'react-dom';
+import { List, ListItem } from 'material-ui/List';
+import FlatButton from 'material-ui/FlatButton';
+import Avatar from 'material-ui/Avatar'
+
+class CommentBox extends React.Component {
+  render() {
+    const { id, comment, username, thumbnail } = this.props.commentObject;
+    console.log('thumbnail: ' + thumbnail);
+    console.log('username: ' + username);
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/commentbox.jsx]: remove console

File: react-redux/contactsMVC/components/commentbox.jsx [hunk 12]
```diff
@@ -0,0 +1,45 @@
+import React, { PropTypes } from 'react';
+import ReactDOM from 'react-dom';
+import { List, ListItem } from 'material-ui/List';
+import FlatButton from 'material-ui/FlatButton';
+import Avatar from 'material-ui/Avatar'
+
+class CommentBox extends React.Component {
+  render() {
+    const { id, comment, username, thumbnail } = this.props.commentObject;
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/commentbox.jsx]: you can do this when all variable names are short. But when they are longer, you may

``` javascript
const {
  id,
  comment,
  username,
  thumbnail,
} = this.props.commentObject
```

File: react-redux/contactsMVC/components/commentbox.jsx [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+            <Avatar src={thumbnail} className="avatar" />
+            {username}
+            <div>
+              {comment}
+            </div>
+          </div>
+          <FlatButton
+            label="Delete"
+            secondary={true}
+            onClick={() => this.props.store.dispatch({ type: 'DEL_COMMENT', id })}
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/commentbox.jsx]: do you know why you don't have to bind here?

File: react-redux/contactsMVC/components/commentbox.jsx [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+          />
+        </div>
+      </ListItem>
+
+    )
+  }
+}
+
+CommentBox.props = {
+  commentObject: PropTypes.object,
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/components/commentbox.jsx]: if it is required, put isRequired here

File: react-redux/contactsMVC/contacts.js [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+        ...state,
+        {
+          id: action.id,
+          username: action.username,
+          thumbnail: action.thumbnail,
+          comment: action.comment
+        }
+      ];
+    case 'DEL_COMMENT':
+      return state.filter( s => s.id !== action.id );
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/contacts.js]: no extra spaces in function call

``` javascript
state.filter(s => s.id !== action.id)
```

File: react-redux/contactsMVC/contacts.js [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+      type: 'ADD_COMMENT'
+    },
+    {
+      id: 1,
+      username: 'hehe',
+      thumnail: 'e e e',
+      comment: 'test 2',
+      type: 'ADD_COMMENT'
+    }
+  ];
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/contacts.js]: remove your test data

File: react-redux/contactsMVC/contacts.js [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+          thumbnail: action.thumbnail,
+          comment: action.comment
+        }
+      ];
+    case 'DEL_COMMENT':
+      return state.filter( s => s.id !== action.id );
+    default:
+      return state;
+  }
+}
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/contacts.js]: put your reducer into another file. you may call it reducer.js in small project.
in large project, you may organise your project in the way introduced here
https://github.com/erikras/ducks-modular-redux

File: react-redux/contactsMVC/contacts.js [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+  ];
+
+injectTapEventPlugin();
+const CommentApp = (props) => (
+  <MuiThemeProvider>
+    <div>
+      <Conversation store={props.store}/>
+    </div>
+  </MuiThemeProvider>
+);
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/contacts.js]: Put these lines of code into App.js and import from that

File: react-redux/contactsMVC/package.json [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+    "react": "^15.3.1",
+    "react-dom": "^15.3.1",
+    "react-tap-event-plugin": "^1.0.0",
+    "react-thunk": "^1.0.0",
+    "redux": "^3.6.0",
+    "redux-logger": "^2.6.1",
+    "redux-thunk": "^2.1.0",
+    "superagent": "^2.2.0",
+    "superagent-as-promised": "^4.0.0",
+    "webpack": "^1.13.2"
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/package.json]: webpack, babel-loader, expect should be in devDependencies

Comment (ZhengHe-MD) [reply react-redux/contactsMVC/package.json]: use 

``` sh
npm install something --save-dev
```

or 

``` sh
npm install something -D
```

File: react-redux/contactsMVC/webpack.config.js [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+  output: {
+    path: __dirname,
+    filename: 'bundle.js'
+  },
+  resolve: {
+    extensions: ["", ".js", ".jsx"]
+  },
+  // devServer: {
+  //   contentBase: path.join(__dirname, 'dist')
+  // },
```
Comment (ZhengHe-MD) [inline react-redux/contactsMVC/webpack.config.js:15]: remove commented code

**Review by ZhengHe-MD — CHANGES_REQUESTED** [review]
You may merge this PR after you fix my comments.
You still need to incorporate
**react-redux**
**redux-thunk**
to complete this project

File: react-redux/contactsMVC/components/Conversation.jsx [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+    this.state = {
+      inputValue: ''
+    }
+  }
+
+  onChangeInputValue (e) {
+    this.setState({ inputValue: e.target.value });
+  }
+
+  getShowComments(showComments) {
```
Comment (maicaiyao) [inline react-redux/contactsMVC/components/Conversation.jsx]: better name? showComments is a verb

File: react-redux/contactsMVC/components/Conversation.jsx [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+          <FlatButton
+            label="Save"
+            primary={true}
+            onClick={ this.onGetRandomUser }
+          />
+        </Card>
+    );
+  }
+
+  onGetRandomUser() {
```
Comment (maicaiyao) [inline react-redux/contactsMVC/components/Conversation.jsx]: react component should be pure. async operations should be in action creators.

**Review by maicaiyao — CHANGES_REQUESTED** [review]

File: react-redux/contactsMVC/components/Conversation.jsx (hunk 21 above)
Comment (ZhengHe-MD) [reply react-redux/contactsMVC/components/Conversation.jsx]: nice catch.
**getCurrentComments**
**getCommentsToShow**
anyone is OK

File: react-redux/contactsMVC/components/Conversation.jsx (hunk 22 above)
Comment (ZhengHe-MD) [reply react-redux/contactsMVC/components/Conversation.jsx]: he will use redux-thunk to do this in next PR

File: react-redux/contactsMVC/components/Conversation.jsx (hunk 21 above)
Comment (ZhengHe-MD) [reply react-redux/contactsMVC/components/Conversation.jsx]: You should do this before you merge

**Review by ZhengHe-MD — APPROVED** [review]

**Pull request merged**